File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        // Method 3 [Morris Inorder Traversal] O(2n), O(1)
        int count = 0;
        int kthSmallestValue = Integer.MIN_VALUE;
        TreeNode current = root;

        while (current != null) {
            if (current.left == null) {
                count++;
                if (count == k) {
                    kthSmallestValue = current.val;
                    break;
                }
                current = current.right;
            } else {
                TreeNode predecessor = findPredecessor(current);
                if (predecessor.right == null) {
                    predecessor.right = current;
                    current = current.left;
                } else {
                    predecessor.right = null;
                    count++;
                    if (count == k) {
                        kthSmallestValue = current.val;
                        break;
                    }
                    current = current.right;
                }
            }
        }
        return kthSmallestValue;
    }

    private TreeNode findPredecessor(TreeNode node) {
        TreeNode predecessor = node.left;
        while (predecessor.right != null && predecessor.right != node) {
            predecessor = predecessor.right;
        }
        return predecessor;
    }
}

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        // Method 1 [BruteForce] O(n + nlogn), O(n)
        // store all nodes values into list and sort list to find given element

        // Method 2 [Inorder Traversal] O(n), O(n)
        int ans = 0, count = 0;
        findKthSmallest(root, k, ans, count);
        return ans;
    }

    void findKthSmallest(TreeNode *root, int k, int &ans, int &count) {
        if(!root) return;
        findKthSmallest(root->left, k, ans, count);
        count++;
        if(count==k) ans = root->val;
        if(count<k) findKthSmallest(root->right, k, ans, count);
    }
};